profiler/src/ProfilerEngine/Datadog.Linux.ApiWrapper/common.c

```diff
@@ -18,3 +24,119 @@ __attribute__((visibility("hidden"))) inline int is_interrupted_by_profiler(int
 {
     return rc == -1L && error_code == EINTR && interrupted_by_profiler != 0;
 }
+
+char* dlerror(void) __attribute__((weak));
+static void* s_libdl_handle = NULL;
+static void* s_libpthread_handle = NULL;
+void* dlsym(void* handle, const char* symbol) __attribute__((weak));
+int pthread_once(pthread_once_t* control, void (*init)(void)) __attribute__((weak));
+
+static __typeof(dlerror)* s_dlerror = &dlerror;
+static __typeof(dlopen)* s_dlopen = NULL;
+void* __libc_dlopen_mode(const char* filename, int flag) __attribute__((weak));
+void* __libc_dlsym(void* handle, const char* symbol) __attribute__((weak));
+
+static void* __dd_dlopen(const char* filename, int flags)
+{
+    // it can be NULL: the first time we enter the function
+    // or if libdl is not loaded yet
+    if (!dlsym)
+    {
+        // if libdl.so is not loaded, use __libc_dlopen_mode
+        s_dlopen = __libc_dlopen_mode;
+    }
+    else
+    {
+        // since we wrap dlopen, we want to call the real dlopen
+        s_dlopen = dlsym(RTLD_NEXT, "dlopen");
+    }
+    if (s_dlopen)
+    {
+        void* ret = s_dlopen(filename, flags);
+        if (!ret && s_dlerror)
+        {
+            fprintf(stderr, "Failed to dlopen %s (%s)\n", filename, s_dlerror());
```

Comment:
I don't think we should write to stdout/stderr?

Reply:
Agreed, definitely not

Reply:
yeah it was for debugging purpose while implementing it

Reply:
✔️